### PR TITLE
How to build a Text Classification model Using BERT and Tensorflow

### LINKED ISSUE #5225: [Machine Learning] How to build a Text Classification model Using BERT and Tensorflow
#### Proposed title of article
[Machine Learning] How to build a Text Classification model Using BERT and Tensorflow

#### Proposed article introduction
Text classification is a subset of machine learning that involves classifying text into predefined categories. Text classification is a fundamental task in natural language processing. Some of the examples of text classification are sentiment analysis, topic labeling, spam detection, and intent detection.

In this tutorial, we will build a spam detection model.  The spam detection model will classify emails as spam or not spam. This is used to detect unsolicited and unwanted emails. We will build this model using BERT and Tensorflow. BERT will be used to generate sentence encoding for all emails and after that, we will use a simple neural network with one drop out layer and one output layer. It's developed by Google.
Finally, we will use Tensorflow to build the neural networks. Tensorflow will build the input and output layers of our machine learning model.

#### Key takeaways 
1. What is text classification and various examples of text classification.
2. What is Bert Model.
3. How to install a Bert model.
4. Exploring Tenserflow functions.
5. Building a spam filtering model Using BERT and Tensorflow.

#### Article quality
This article is unique since we will be building a unique model. The model will be a spam filtering model which is a type of text classification. We will also use Bert to generate sentence encoding for all emails and after that, we will use a simple neural network with one drop out layer and one output layer. We will also implement dropout regularization to handle overfitting during model building.

#### References
N/A

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by Bravinsimiyu**
### Article checklist

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #5225

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by Kayere** [comment]
Plagiarism check done.

**Comment by hectorkambow** [comment]
@section-engineering-education/content-moderator-team  

This author has expressed interest in doing a potential test review. 

@Bravinsimiyu 

We will keep an eye on this pull request to see if it needs little or no revisions.

File: content/articles/how-to-build-a-text-classification-model-using-bert-and-tensorflow/index.md [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+
+For a reader to understand this tutorial easily, they should:
+- Know how to work with [deep learning models](/engineering-education/building-a-deep-learning-app-using-python/)
+- Know how to use [Pandas](https://pandas.pydata.org/) and [Numpy](https://numpy.org/) in data analysis.
+- Know how to work with [TensorFlow](https://www.tensorflow.org/)
+- Have [Google Colab notebook](https://research.google.com/colaboratory/). We will use Google Colab because it's faster.
+
+### Importing important packages
+
+Let's import the important packages as follows:
```
Comment (Bayler) [inline content/articles/how-to-build-a-text-classification-model-using-bert-and-tensorflow/index.md:54]:
```suggestion
Let's import the required packages as follows:
```

File: content/articles/how-to-build-a-text-classification-model-using-bert-and-tensorflow/index.md [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+author: bravin-wasike
+date: 2021-12-17T00:00:00-17:00
+topics: [Machine Learning]
+excerpt_separator: <!--more-->
+images:
+
+  - url: /engineering-education/how-to-build-a-text-classification-model-using-bert-and-tensorflow/hero.jpg
+    alt: Text Classification model Using BERT and Tensorflow Hero Image
+---
+Text classification is a subset of machine learning that classifies text into predefined categories. Text classification is one of the impotant tasks in natural language processing(NLP).
```
Comment (Bayler) [inline content/articles/how-to-build-a-text-classification-model-using-bert-and-tensorflow/index.md:17]:
```suggestion
Text classification is a subset of machine learning that classifies text into predefined categories. Text classification is one of the important tasks in natural language processing (NLP). 
```

File: content/articles/how-to-build-a-text-classification-model-using-bert-and-tensorflow/index.md [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+This is the machine learning package used to build the neural network. It will create the input and output layers of our machine learning model.
+
+- `tensorflow_hub`:
+This contains a pre-trained machine model used to build our text classification. Our pre-trained model is BERT. We will re-use the BERT model and fine-tune it to meet our needs.
+
+- `tensorflow_text`:
+This will allow us to work with text. In this tutorial, we are solving a text-classification problem.
+
+- `pandas`:
+We will use Pandas to load our dataset. We also use Pandas for data manipulation and analysis. This gives us a clear overview of how our dataset is structured.
```
Comment (Bayler) [inline content/articles/how-to-build-a-text-classification-model-using-bert-and-tensorflow/index.md:74]: Replace `This` in these sentences with `It`

File: content/articles/how-to-build-a-text-classification-model-using-bert-and-tensorflow/index.md [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+
+### Initializing the neural network layers
+
+```python
+l = tf.keras.layers.Dropout(0.1, name="dropout")(outputs['pooled_output'])
+l = tf.keras.layers.Dense(1, activation='sigmoid', name="output")(l)
+```
+The neural network has two layers, the `Dropout` layer, and the `Dense` layer.
+
+- `Droupout` layer
```
Comment (Bayler) [inline content/articles/how-to-build-a-text-classification-model-using-bert-and-tensorflow/index.md:273]:
```suggestion
- `Dropout` layer
```

File: content/articles/how-to-build-a-text-classification-model-using-bert-and-tensorflow/index.md [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+
+- `tensorflow_text`:
+This will allow us to work with text. In this tutorial, we are solving a text-classification problem.
+
+- `pandas`:
+We will use Pandas to load our dataset. We also use Pandas for data manipulation and analysis. This gives us a clear overview of how our dataset is structured.
+
+Now, let's load and explore the dataset we will use in this tutorial. Before we load the dataset, make sure you download this dataset from [here](https://drive.google.com/file/d/1QXzFte4FulQXmgUHBAk8y6IqcCYy8An-/view?usp=sharing)
+
+Now, let's load the dataset that you have downloaded from the link above. Run this command to load the dataset.
```
Comment (Bayler) [inline content/articles/how-to-build-a-text-classification-model-using-bert-and-tensorflow/index.md:78]:
```suggestion
Now, let's load the dataset downloaded using the link above. Run this command to load the dataset.
```

File: content/articles/how-to-build-a-text-classification-model-using-bert-and-tensorflow/index.md [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+
+```python
+747/4825
+```
+```bash
+0.15481865284974095
+```
+This implies that 15% is the spam emails and 85% the ham emails. This indicates class imbalance.
+
+We need to balance the two classes so that to reduce bias during model training.
```
Comment (Bayler) [inline content/articles/how-to-build-a-text-classification-model-using-bert-and-tensorflow/index.md:117]:
```suggestion
We need to balance the two classes to reduce bias during model training.
```

File: content/articles/how-to-build-a-text-classification-model-using-bert-and-tensorflow/index.md [hunk 7]
```diff
@@ -0,0 +1,435 @@
...
+
+### Building model using TensorFlow
+
+There are two types of models that you can build in TensorFlow. Sequential model and a functional model.
+
+In a sequential model, layers are built on top of each other, layer by layer. In a sequential model, we don't have multiple inputs and outputs.
+
+Functional models are more robust and flexible. They do not create layers in sequential order. In the functional model, we have multiple inputs and outputs.
+
+In this tutorial, we will use the functional approach to build our model.
```
Comment (Bayler) [inline content/articles/how-to-build-a-text-classification-model-using-bert-and-tensorflow/index.md:250]:
```suggestion
This tutorial will use the functional approach to build our model.
```

File: content/articles/how-to-build-a-text-classification-model-using-bert-and-tensorflow/index.md [hunk 8]
```diff
@@ -0,0 +1,435 @@
...
+
+![Model summary](/engineering-education/how-to-build-a-text-classification-model-using-bert-and-tensorflow/model-summary.png)
+
+The image above shows all the input and output layers we have initialized for our model. The output also shows the total params, trainable params, and non-trainable params.
+
+- Total params:
+This represents all the parameters in our model.
+
+- Trainable params:
+This represents the parameters that we will train.
```
Comment (Bayler) [inline content/articles/how-to-build-a-text-classification-model-using-bert-and-tensorflow/index.md:307]: Replace `This` with `It`.

**Comment by Bayler** [comment]
@Bravinsimiyu Please work on the minor edits to improve on flow before we can move on to final review.

**Comment by Bravinsimiyu** [comment]
Hello @Bayler,
I have made the requested changes, kindly review
Thanks

**Comment by Bayler** [comment]
@Bravinsimiyu Your tutorial is easy to follow. You've explained methods and functions used in detail. That's good! Especially for a beginner. Keep it up! Just take note of the commits I've made. Otherwise, satisfied with the changes as it is.

**Comment by hectorkambow** [comment]
jumping on this one
